File: project_euler/Problem092/Problem092.java
package project_euler;

import java.util.HashSet;
import java.util.Set;

public class Problem092 {
    public void solution() {
        int target = 10000000;
        int border = 9 * 9 * (int) Math.log10(target);
        Set<Integer> chains = new HashSet<>();
        int pivot = 89;
        chains.add(pivot);

        int[] squares = new int[10];
        for (int i = 0; i < 10; ++i) {
            squares[i] = i * i;
        }

        int result = 0;
        for (int i = 1; i <= border; ++i) {
            if (getChains(i, chains, pivot, squares)) {
                ++result;
            }
        }
        for (int i = border + 1; i < target; ++i) {
            if (chains.contains(squareDigits(i, squares))) {
                ++result;
            }
        }
        System.out.println(result);
    }

    private boolean getChains(int value, Set<Integer> chains, int pivot, int[] squares) {
        if (chains.contains(value) || pivot == value) {
            return true;
        }
        if (1 == value) {
            return false;
        }

        boolean result = getChains(squareDigits(value, squares), chains, pivot, squares);
        if (result) {
            chains.add(value);
        }
        return result;
    }

    private int squareDigits(int value, int[] squares) {
        int result = 0;
        while (0 != value) {
            result += squares[value % 10];
            value /= 10;
        }
        return result;
    }
}

File: project_euler/Problem092/Problem092.py
import math
from typing import List

class Problem092:
    def solution(self) -> None:
        target = 10000000
        border, chains, pivot = 9 * 9 * int(math.log10(target)), set(), 89
        chains.add(pivot)

        squares = [i * i for i in range(10)]
        result = 0
        for i in range(1, border + 1):
            if self.__get_chains(i, chains, pivot, squares):
                result += 1
        for i in range(border + 1, target):
            if self.__square_digits(i, squares) in chains:
                result += 1
        print(result)

    def __get_chains(self, value: int, chains: set[int], pivot: int, squares: List[int]) -> bool:
        if value in chains or pivot == value:
            return True
        if 1 == value:
            return False

        result = self.__get_chains(self.__square_digits(value, squares), chains, pivot, squares)
        if result:
            chains.add(value)
        return result

    def __square_digits(self, value: int, squares: List[int]) -> int:
        result = 0
        while value:
            result += squares[value % 10]
            value //= 10
        return result
